src/shared/bootspec.c

```diff
@@ -755,41 +755,88 @@ static int boot_entry_load_unified(
+                PeHeader **pe_header) {
 
-        _cleanup_free_ IMAGE_SECTION_HEADER *sections = NULL;
         _cleanup_free_ IMAGE_DOS_HEADER *dos_header = NULL;
-        _cleanup_free_ char *osrel = NULL, *cmdline = NULL;
-        _cleanup_free_ PeHeader *pe_header = NULL;
         int r;
 
         assert(fd >= 0);
         assert(path);
 
-        r = pe_load_headers(fd, &dos_header, &pe_header);
+        r = pe_load_headers(fd, &dos_header, pe_header);
         if (r < 0)
                 return log_warning_errno(r, "Failed to parse PE file '%s': %m", path);
 
-        r = pe_load_sections(fd, dos_header, pe_header, &sections);
+        r = pe_load_sections(fd, dos_header, *pe_header, sections);
         if (r < 0)
                 return log_warning_errno(r, "Failed to parse PE sections of '%s': %m", path);
 
-        if (!pe_is_uki(pe_header, sections))
-                return log_warning_errno(SYNTHETIC_ERRNO(EBADMSG), "Parsed PE file '%s' is not a UKI.", path);
+        return 0;
+}
 
-        r = pe_read_section_data(fd, pe_header, sections, ".osrel", PE_SECTION_SIZE_MAX, (void**) &osrel, NULL);
-        if (r < 0)
-                return log_warning_errno(r, "Failed to read .osrel section of '%s': %m", path);
+static int find_cmdline_section(
+                int fd,
+                const char *path,
+                IMAGE_SECTION_HEADER *sections,
+                PeHeader *pe_header,
+                char **ret_cmdline) {
 
-        r = pe_read_section_data(fd, pe_header, sections, ".cmdline", PE_SECTION_SIZE_MAX, (void**) &cmdline, NULL);
+        int r;
+
+        if (!ret_cmdline)
+                return 0;
+
+        r = pe_read_section_data(fd, pe_header, sections, ".cmdline", PE_SECTION_SIZE_MAX, (void**) ret_cmdline, NULL);
         if (r < 0 && r != -ENXIO) /* cmdline is optional */
                 return log_warning_errno(r, "Failed to read .cmdline section of '%s': %m", path);
 
-        if (ret_osrelease)
-                *ret_osrelease = TAKE_PTR(osrel);
-        if (ret_cmdline)
-                *ret_cmdline = TAKE_PTR(cmdline);
+        return 0;
+}
+
+static int find_osrel_section(
+                int fd,
+                const char *path,
+                IMAGE_SECTION_HEADER *sections,
+                PeHeader *pe_header,
+                char **ret_osrelease) {
+
+        int r;
+
+        if (!ret_osrelease)
+                return 0;
+
+        r = pe_read_section_data(fd, pe_header, sections, ".osrel", PE_SECTION_SIZE_MAX, (void**) ret_osrelease, NULL);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to read .osrel section of '%s': %m", path);
+
+        return 0;
+}
+
+static int find_uki_sections(
+                int fd,
+                const char *path,
+                char **ret_osrelease,
+                char **ret_cmdline) {
+
+        _cleanup_free_ IMAGE_SECTION_HEADER *sections = NULL;
+        _cleanup_free_ PeHeader *pe_header = NULL;
```

Comment:
Why not just make this a local variable `PeHeader pe_header` ?

Reply:
pe_header is allocated in `pe_load_headers`, so unless I am missing something it makes no sense removing the pointer and using a variable.